Sources/CancellableThenable.swift

```diff
@@ -0,0 +1,520 @@
+import Dispatch
+
+/**
+ CancellableThenable represents an asynchronous operation that can be both chained and cancelled.  When chained, all CancellableThenable members of the chain are cancelled when `cancel` is called on the associated CancelContext.
+ */
+public protocol CancellableThenable: class {
+    /// Type of the delegate `thenable`
+    associatedtype U: Thenable
+    
+    /// Delegate `thenable` for this `CancellableThenable`
+    var thenable: U { get }
+
+    /// The `CancelContext` associated with this `CancellableThenable`
+    var cancelContext: CancelContext { get }
+    
+    /// Tracks the cancel items for this `CancellableThenable`.  These items are removed from the associated `CancelContext` when the thenable resolves.
+    var cancelItemList: CancelItemList { get }
+}
+
+public extension CancellableThenable {
+    /// Append the `task` and `reject` function for a cancellable task to the cancel context
+    func appendCancellableTask(_ task: CancellableTask?, reject: ((Error) -> Void)?) {
+        self.cancelContext.append(task: task, reject: reject, thenable: self)
+    }
+    
+    /// Append the cancel context associated with `from` to our cancel context.  Typically `from` is a branch of our chain.
+    func appendCancelContext<Z: CancellableThenable>(from: Z) {
+        self.cancelContext.append(context: from.cancelContext, thenable: self)
+    }
+    
+    /**
+     Cancel all members of the promise chain and their associated asynchronous operations.
+
+     - Parameter error: Specifies the cancellation error to use for the cancel operation, defaults to `PMKError.cancelled`
+     */
+    func cancel(error: Error? = nil) {
```

Comment:
I'd prefer `cancel(with error: Error)` also seems like it should be a `CancellableError`, the default should be `PMKError.cancelled` and not `nil`.

Reply:
Great idea!  Will do.

Reply:
Changed signature to `cancel(with error: Error)`.

Checked in!